File: ws362_pset04/pset04.py
""" Problem Set 04 starter code

Please make sure your code runs on Python version 3.5.0

Due date: 2016-03-04 13:00
"""

import numpy as np
import scipy.optimize

def dualFun(alphas, X, y):
    res = -sum(alphas)
    alphas = np.array([alphas])
    y = np.array([y])
    n = X.shape[0]
    k = 0
    res += 1.0 / 2.0 * np.dot(np.dot(alphas, np.dot(np.dot(y.T, y), np.dot(X, X.T))), alphas.T)
    return res

def my_dual_svm(X, y, C=1):
    bnd = ((0, C),) * X.shape[0]
    alphas = np.array([0] * X.shape[0])
    alphas = scipy.optimize.minimize(dualFun,alphas,args = (X,y),method='L-BFGS-B', options={'disp': True},bounds=bnd).x
    betas = np.zeros((X.shape[1]))
    for i in range(X.shape[0]):
        betas += alphas[i] * y[i] * X[i]
    return betas


def my_primal_svm(X, y, lam=1, k=5, T=100):
    n = X.shape[1]
    num = X.shape[0]
    w1 = np.random.normal(0, 1.0, n)
    w1 = w1 * np.linalg.norm(w1) / np.sqrt(lam)
    w_t = w1
    for t in range(1, T + 1):
        A_t_idx = np.random.choice(range(0, num), k)
        A_t_plus_idx = A_t_idx[(np.dot(w_t, X[A_t_idx].T) < 1)]
        eta_t = 1.0 / (t * lam)
        w_t_half = (1 - eta_t * lam) * w_t + eta_t / k * np.sum((np.dot(X[A_t_plus_idx].T, y[A_t_plus_idx])))
        w_t_half_norm = np.linalg.norm(w_t_half)
        tmp = None
        if w_t_half_norm == 0.0:
            tmp = 1
        else:
            tmp = 1.0/np.sqrt(lam)/np.linalg.norm(w_t_half)
        # print("2:",np.linalg.norm(w_t_half))
        w_t_plus_1 = min(1, tmp) * w_t_half
        w_t = w_t_plus_1
    # print("here",w_t)
    return w_t

File: ws362_pset04/pset04.R
# Problem Set 04 starter code
#
# Please make sure your code runs on R version 3.2.2
# Due date: 2016-03-04 13:00

#' Support vector machine - Dual problem
#'
#' SVM classification for a numeric test matrix. The
#'   returned result is the vector of coefficients from
#'   the support vector machine (beta, *not* alpha!).
#'
#' @param X     an n by p numeric matrix; the data
#'                matrix of predictors
#' @param y     a length n numberic vector; the observed
#'                response
#' @param C     positive numeric value giving the cost of
#'                parameter in the support vector machine
#'
#' @return      a length p vector giving the coefficients of
#'                beta in the SVM
dualFunc <- function(alphas, Xs, ys) {
  print(class(alphas))
  res = -sum(alphas)
  for (i in 1:nrow(Xs)) {
    for (j in 1:nrow(Xs)){
      res = res + 1/ 2 * alphas[i] * alphas[j] * y[i] *y[j] * ((Xs[i] %*% Xs[j])[1,1])
    }}
  res
}
my_dual_svm <- function(X, y, C=1L) {
  alphas = c(rep(C/2,nrow(X))) # this is the right dimension for the output
  alphas = optim(par = alphas, fn = dualFunc, Xs = X, ys = y, method="L-BFGS-B", 
                 lower=rep(0, nrow(X)), upper=rep(C, nrow(X)))
  alphas
}

#' Support vector machine - Primal problem
#'
#' SVM classification for a numeric test matrix. The
#'   returned result is the vector of coefficients from
#'   the support vector machine (beta, *not* alpha!).
#'   The algorithm solves the primal problem, written
#'   as a penalized hinge loss, using the Pegasos algorithm
#'
#' @param X       an n by p numeric matrix; the data
#'                  matrix of predictors
#' @param y       a length n numberic vector; the observed
#'                  response
#' @param lam     positive numeric value giving the tuning
#'                  parameter in the (primal, penalized format)
#'                  of the support vector machine
#' @param k       positive integer giving the number of samples
#'                  selected in each iteration of the algorithm
#' @param T       positive integer giving the total number of
#'                   iterations to run
#'
#' @return      a length p vector giving the coefficients of
#'                beta in the SVM
my_primal_svm <- function(X, y, lam=1, k=5, T=100) {
  rep(0,ncol(X)) # this is the right dimension for the output
}



